src/libraries/Common/src/System/Net/ArrayBuffer.cs

```diff
@@ -30,8 +31,12 @@ internal struct ArrayBuffer : IDisposable
 
         public ArrayBuffer(int initialSize, bool usePool = false)
         {
+            Debug.Assert(initialSize > 0 || usePool);
+
             _usePool = usePool;
-            _bytes = usePool ? ArrayPool<byte>.Shared.Rent(initialSize) : new byte[initialSize];
+            _bytes = initialSize == 0
+                ? Array.Empty<byte>()
+                : usePool ? ArrayPool<byte>.Shared.Rent(initialSize) : new byte[initialSize];
```

Comment:
Note ArrayPool will use Array.Empty for 0 byte requests, so this could also be:
```C#
_bytes =
    usePool ? ArrayPool<byte>.Shared.Rent(initialSize) :
    initialSize == 0 ? Array.Empty<byte>() :
    new byte[initialSize];
```
It's unlikely to really matter, but if we expect initialSize to most commonly be non-zero, you might want to reorder it.